File: plugins/airsstack-plugin-dev/hooks/cache_sync.py
#!/usr/bin/env python3
"""airsstack-plugin-dev — PostToolUse cache-sync hook.

Mirrors an edited plugins/<plugin>/<rel> file into that plugin's install cache
(~/.claude/plugins/cache/<marketplace>/<plugin>/<version>/<rel>) so plugin
development at a fixed version needs no manual cp and no reinstall.

Only plugins installed from the `airsstack` marketplace are touched. Every
failure mode no-ops to exit 0; the hook never blocks the triggering tool.
"""

import json
import os
import shutil
import sys

MARKETPLACE = "airsstack"
CACHE_ROOT = os.path.join(
    os.path.expanduser("~"), ".claude", "plugins", "cache"
)
INSTALLED_PLUGINS = os.path.join(
    os.path.expanduser("~"), ".claude", "plugins", "installed_plugins.json"
)
DEBUG = bool(os.environ.get("AIRSSTACK_PLUGIN_DEV_DEBUG"))


def _debug(msg):
    if DEBUG:
        sys.stderr.write("[cache-sync] " + msg + "\n")


def extract_plugin_rel(path):
    """Return (plugin, rel) for a path under plugins/<plugin>/, else None.

    `rel` is the path remainder after plugins/<plugin>/. Returns None when the
    path has no plugins/<plugin>/ segment or names a plugin dir with no file
    remainder.
    """
    parts = path.split(os.sep)
    for i, part in enumerate(parts):
        if part == "plugins" and i + 2 < len(parts):
            plugin = parts[i + 1]
            rel_parts = parts[i + 2:]
            if not plugin or not rel_parts:
                return None
            return plugin, os.sep.join(rel_parts)
    return None


def resolve_install_paths(installed_data, plugin):
    """Distinct installPath values for `<plugin>@airsstack`, first-seen order.

    The `@airsstack` suffix is the marketplace gate: plugins from any other
    marketplace are never selected.
    """
    key = plugin + "@" + MARKETPLACE
    entries = (installed_data.get("plugins") or {}).get(key) or []
    seen = []
    for entry in entries:
        ip = entry.get("installPath")
        if ip and ip not in seen:
            seen.append(ip)
    return seen


def is_within(child, parent):
    """True if `child` is `parent` or nested under it (normalized paths)."""
    parent_n = os.path.normpath(parent)
    child_n = os.path.normpath(child)
    return child_n == parent_n or child_n.startswith(parent_n + os.sep)

File: plugins/airsstack-plugin-dev/hooks/cache_sync.lua
-- airsstack-plugin-dev — PostToolUse cache-sync hook.
--
-- Mirrors an edited plugins/<plugin>/<rel> file into that plugin's install cache
-- (~/.claude/plugins/cache/<marketplace>/<plugin>/<version>/<rel>) so plugin development at a
-- fixed version needs no manual cp and no reinstall.
--
-- Only plugins installed from the `airsstack` marketplace are touched. Every failure mode no-ops;
-- the hook never blocks the triggering tool.
--
--   airsl run --fail-open --policy confined \
--     --allow-env HOME --allow-env AIRSSTACK_PLUGIN_DEV_DEBUG \
--     --allow-read / --allow-write "$HOME/.claude/plugins/cache" \
--     hooks/cache_sync.lua

local cache = require("lib.cache")
local path = airsstack.path

local home = airsstack.env.get("HOME") or ""
local debugging = airsstack.env.get("AIRSSTACK_PLUGIN_DEV_DEBUG")

local function note(message)
  if debugging and debugging ~= "" then
    airsstack.stdio.error("[cache-sync] " .. message .. "\n")
  end
end

local ok, payload = pcall(airsstack.hook.payload)
if not ok or type(payload) ~= "table" then
  return
end

local tool_input = type(payload.tool_input) == "table" and payload.tool_input or {}
local source = tool_input.file_path
if type(source) ~= "string" or source == "" then
  return
end

source = path.absolute(source)
if not cache.is_file(source) then
  return
end

local plugin, rel = cache.extract_plugin_rel(source)
if not plugin then
  return
end

local registry = cache.load_registry(cache.registry_path(home))
local root = cache.cache_root(home)

for _, install_path in ipairs(cache.install_paths(registry, plugin)) do
  local dest, reason = cache.sync_one(source, rel, install_path, root)
  if dest then
    note("synced: " .. source .. " -> " .. dest)
  else
    note("skip (" .. reason .. "): " .. path.join(install_path, rel))
  end
end
